File: frontend/src/features/adjustments/api/adjustments.ts
import { apiFetch } from '../../../lib/api'

export type AdjustmentStatus = 'draft' | 'confirmed' | 'cancelled'
export type StockDirection = 'in' | 'out'
export type AdjustmentReason = 'inventory_count' | 'damage' | 'loss' | 'expired' | 'correction' | 'other'

export interface AdjustmentListItem {
  id: string
  adjustment_number: string
  warehouse_id: string
  adjustment_date: string
  reason: AdjustmentReason
  status: AdjustmentStatus
  notes: string | null
  created_at: string
  updated_at: string
  created_by_user_id: string | null
  updated_by_user_id: string | null
}

export interface AdjustmentListOut {
  items: AdjustmentListItem[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface AdjustmentListParams {
  warehouse_id?: string
  status?: AdjustmentStatus
  date_from?: string
  date_to?: string
  page?: number
  page_size?: number
}

export interface AdjustmentItemOut {
  id: string
  adjustment_id: string
  product_id: string
  product_unit_id: string
  quantity: string
  quantity_base: string
  direction: StockDirection
  unit_cost_base: string | null
  notes: string | null
}

export interface AdjustmentOut extends AdjustmentListItem {
  items: AdjustmentItemOut[]
}

export interface AdjustmentCreate {
  id: string
  warehouse_id: string
  adjustment_date: string
  reason: AdjustmentReason
  notes?: string | null
}

export interface AdjustmentUpdate {
  warehouse_id?: string
  adjustment_date?: string
  reason?: AdjustmentReason
  notes?: string | null
}

export interface AdjustmentItemCreate {
  id: string
  product_id: string
  product_unit_id: string
  quantity: number
  direction: StockDirection
  unit_cost_base?: number | null
  notes?: string | null
}

export interface AdjustmentAuditEntry {
  id: string
  action: 'create' | 'confirm' | 'cancel'
  user_id: string
  user_name: string
  created_at: string
  changes: Record<string, unknown> | null
}

export function fetchAdjustment(id: string): Promise<AdjustmentOut> {
  return apiFetch<AdjustmentOut>(`/adjustments/${id}`)
}

export function fetchAdjustments(params: AdjustmentListParams = {}): Promise<AdjustmentListOut> {
  const qs = new URLSearchParams()
  if (params.warehouse_id) qs.set('warehouse_id', params.warehouse_id)
  if (params.status) qs.set('status', params.status)
  if (params.date_from) qs.set('date_from', params.date_from)
  if (params.date_to) qs.set('date_to', params.date_to)
  if (params.page != null) qs.set('page', String(params.page))
  if (params.page_size != null) qs.set('page_size', String(params.page_size))
  const query = qs.toString()
  return apiFetch<AdjustmentListOut>(`/adjustments${query ? `?${query}` : ''}`)
}

export function createAdjustment(data: AdjustmentCreate): Promise<AdjustmentOut> {
  return apiFetch<AdjustmentOut>('/adjustments', { method: 'POST', body: JSON.stringify(data) })
}

export function updateAdjustment(id: string, data: AdjustmentUpdate): Promise<AdjustmentOut> {
  return apiFetch<AdjustmentOut>(`/adjustments/${id}`, { method: 'PATCH', body: JSON.stringify(data) })
}

export function deleteAdjustment(id: string): Promise<void> {
  return apiFetch<void>(`/adjustments/${id}`, { method: 'DELETE' })
}

export function addAdjustmentItem(adjustmentId: string, data: AdjustmentItemCreate): Promise<AdjustmentItemOut> {
  return apiFetch<AdjustmentItemOut>(`/adjustments/${adjustmentId}/items`, {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function removeAdjustmentItem(adjustmentId: string, itemId: string): Promise<void> {
  return apiFetch<void>(`/adjustments/${adjustmentId}/items/${itemId}`, { method: 'DELETE' })
}

export function confirmAdjustment(id: string): Promise<AdjustmentOut> {
  return apiFetch<AdjustmentOut>(`/adjustments/${id}/confirm`, { method: 'POST' })
}

export function cancelAdjustment(id: string, reason: string): Promise<AdjustmentOut> {
  return apiFetch<AdjustmentOut>(`/adjustments/${id}/cancel`, {
    method: 'POST',
    body: JSON.stringify({ reason }),
  })
}

export function fetchAdjustmentAudit(id: string): Promise<AdjustmentAuditEntry[]> {
  return apiFetch<AdjustmentAuditEntry[]>(`/adjustments/${id}/audit`)
}

File: backend/app/schemas/adjustments.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, model_validator

from app.enums import AdjustmentReason, AdjustmentStatus, AuditAction, StockDirection


class AdjustmentCreate(BaseModel):
    id: UUID
    warehouse_id: UUID
    adjustment_date: date
    reason: AdjustmentReason
    notes: str | None = None


class AdjustmentUpdate(BaseModel):
    warehouse_id: UUID | None = None
    adjustment_date: date | None = None
    reason: AdjustmentReason | None = None
    notes: str | None = None


class AdjustmentCancelBody(BaseModel):
    reason: str


class AdjustmentItemCreate(BaseModel):
    id: UUID
    product_id: UUID
    product_unit_id: UUID
    quantity: Decimal
    direction: StockDirection
    unit_cost_base: Decimal | None = None
    notes: str | None = None

    @model_validator(mode="after")
    def cost_required_for_in(self) -> "AdjustmentItemCreate":
        if self.direction == StockDirection.IN and self.unit_cost_base is None:
            raise ValueError("unit_cost_base requerido cuando direction='in'")
        return self


class AdjustmentItemUpdate(BaseModel):
    quantity: Decimal | None = None
    direction: StockDirection | None = None
    unit_cost_base: Decimal | None = None
    notes: str | None = None


class AdjustmentItemOut(BaseModel):
    id: UUID
    adjustment_id: UUID
    product_id: UUID
    product_unit_id: UUID
    quantity: Decimal
    quantity_base: Decimal
    direction: StockDirection
    unit_cost_base: Decimal | None = None
    notes: str | None = None


class AdjustmentListItem(BaseModel):
    id: UUID
    adjustment_number: str
    warehouse_id: UUID
    adjustment_date: date
    reason: AdjustmentReason
    status: AdjustmentStatus
    notes: str | None = None
    created_at: datetime
    updated_at: datetime
    created_by_user_id: UUID | None = None
    updated_by_user_id: UUID | None = None


class AdjustmentListOut(BaseModel):
    items: list[AdjustmentListItem]
    total: int
    page: int
    page_size: int
    total_pages: int


class AdjustmentOut(AdjustmentListItem):
    items: list[AdjustmentItemOut]


class AdjustmentAuditEntry(BaseModel):
    id: UUID
    action: AuditAction
    user_id: UUID
    user_name: str
    created_at: datetime
    changes: dict | None = None
